// moongose/firebase_config.js
var admin = require("firebase-admin");
var serviceAccount = {
  "type": "service_account",
  "project_id": "dynamic-tide-315303",
  "private_key_id": "e5464e145715de7351cc8ac619c3f4cf61d20ee6",
  "private_key": "-----BEGIN PRIVATE KEY-----\nMIIEvwIBADANBgkqhkiG9w0BAQEFAASCBKkwggSlAgEAAoIBAQC4Jf8cT9HVBLN8\nY5f8gfMfkRaPgo8tDTUtsdrnnHRd/9mjVWgrhSSuJ4f4aV1DxjWQ7jgjo7SKDQXj\n2vfGOeuJsvEZDFnjW7Qmtso5s9LDYNydQFR51YDxFkON23cmOEEkGkR4V4XKZJTB\n/8xAsqoTYSqtCX2Dtr7wFVXivwxFdR3G1SVogpMLayNAaesljlb9KNExOBAy35xW\nnTjBnw8RcZrJ0EXiwg0xmwg8YR5UefOOGe8sJOmhWGYNlAX22CkZF1zUDxBKHEuX\nlCufQUJ49okyZ2AHM7C5pkZP0YzsVGbxDP/SDvsgFF4gvm0X/HlmB4ojYJlK8gz+\nEmTkv0FdAgMBAAECggEAAgVeu6GPoLocuyp/pkSJ1ciamggZiKnKsd3wFFcl+sse\nSRncmw+qw6OI7TvAy4ffGiBXKuL2zgYPEL1cN0d284usdDUrvmBJBZy/cusrjrvf\nJ/WnB7BFaj/vjVzBfcCjzGDia/yv6fT5Ifjr9bDQR4arOHtwbnPgvBNL3kFxAcke\nsYDEDZsYQUpWEi5RwE433JOikocazw9zKez/xcoNp9xTXMENONOuajOmc7agKIPk\nrnZJeX4udLzNb2Dxscl8rGtnnHbP80XuD1P6TT6e5qa2Cf/gLkY1C6MpjgafLFPp\nTOrsqE1O4PcY1SeuFbu0uq/JYS+eseIMzMT44qkL6QKBgQD7j/jK6KKjDtHEx2DW\ne82Gt1yGB50t94lQNAPrSfCOdC7uRw9aZfT1GXNXKN3ZQveoQiN61GSEX9xm68VN\n2kEm8wuj3mTzApLD2hBAbnmqorutQVQ4Mp5u4ppqwG/v50nFIt15IlDv83HXoD/p\nQe+D2B9YXVk4nno5AJO+xAEjVQKBgQC7ZZcx4W9WFue43zmwVh8TFPLFAmPcvsKC\nGi01Vv+Vjjqs1prYIfxB3ks6x36wPbbmtcXkoHuhhnErHQulQRjB7+rJ0uZip5dK\n9ooDIu0XbbnEpH31UxDT6gOxI0t5n3vFXrNevf3eOvgyl3JqxjKF1OidoQM85u5J\nVQtzoM+16QKBgQCpEs+FGKvOabRkzBfyr8ynwW8nCdEDRv578Cm+JrjOFZpBPXM9\nL49eeBeEOgw47IBlDuantD5GleMdLVMziwAE/I8TOPKiPVWE2nfXxihxneQDgwfm\nqWbAYf6kxue01r9HBrLRqlW5n+YgpSyvSkh/VW9jxYGkwyakTxX9veCB7QKBgQCi\n70O0lQQO4D992Gj6TwcZuj8S2G8ngZzxLf1cwVXcHRwbJ0Te4z3Hr5XNd6NJbX9X\nvXhgvsB/wrsm90wm1SDHjuifroWuBXVgMlfxPi7TDg5VUYhLA3xAJNos7Dk1eFtP\nUnW2zI6s6KzgoBlViRpojqPqWhLMoyVQEWexk4LF8QKBgQDTkrVTCMrtqJvbF9bq\nUrkOAoAupeKoELKkCK901uNs9YkmEgZUc7r6fE/oCSeAszcD6WLgNlDARvI+yxiY\nTZTjNXUdTi6SzPUQP46jbQtKcE3dRzSVkG6BcUSSMxiFLKM7u0VDVTqaadTo7BDO\nKY0vhbvnyx+y2y2lvLPHlBUUAA==\n-----END PRIVATE KEY-----\n",
  "client_email": "firebase-adminsdk-cfu7y@dynamic-tide-315303.iam.gserviceaccount.com",
  "client_id": "115147210186055903132",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/firebase-adminsdk-cfu7y%40dynamic-tide-315303.iam.gserviceaccount.com"
}
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});
const sendFirebaseNotification = async (data, title, body) => {
  var payload = {
        notification: {
            title: title,
            body: body,
        }
    };
    var options = {
        priority: "high",
        timeToLive: 60 * 60 *24
    };
    admin.messaging().sendToDevice(data, payload, options)
    .then(function(response) {
        console.log(response);
    })
    .catch(function(error) {
        console.log(error);
    });
}
module.exports = {
    sendFirebaseNotification
}
